Remove Blogging reminders feature config

ravishanker opened the pull request:
This PR 
- Updates Blogging Reminder feature to shown only on Jetpack 
- removes the `BloggingReminderFeatureConfig`, since it has been released over an year ago.
- Updates Site Settings to reflect the change

The Blogging Reminders feature has two main UI components:

A one-time prompt shown per-blog to promote the feature.
A setting in Site Settings for the user to adjust when the local reminder notification is shown.

To test:

The Blogging Reminders feature has two main UI components:

A one-time prompt shown per-blog to promote the feature.
A setting in Site Settings for the user to adjust when the local reminder notification is shown.

JP app
- Add a new site to the Jetpack app
- Publish a new Post (FAB -> Blog post)
- Confirm that the Blogging Reminders prompt is shown
- Additionally, navigate to Menu -> Site Settings -> Verify Reminders setting is visible under **Blogging** heading, and works

WP app
- Add a site to WP app
- Publish a Post (FAB -> Blog post
- Confirm that the Blogging Reminders prompt is not shown
- Additionally, navigate to Menu -> Site Settings -> Verify no Reminders setting is shown


## Regression Notes
1. Potential unintended areas of impact
Tested on both WP and JP apps

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing, and unit tests

3. What automated tests I added (or what prevented me from doing so)
Updated unit tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ravishanker committed:
Show blogging reminders only on Jetpack app

Show blogging reminders only on Jetpack app, and if jetpack connected

ravishanker committed:
Remove feature config

Remove BloggingRemindersFeatureConfig

ravishanker committed:
Update BloggingRemindersManagerTest.kt

ravishanker committed:
Merge branch 'trunk' into Blogging-Reminders-Don't-show-prompt-if-not-allowed-for-blog

ravishanker committed:
Update BloggingRemindersViewModelTest.kt

ravishanker committed:
Update BloggingRemindersManagerTest.kt

ravishanker committed:
Update BloggingRemindersManagerTest.kt

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsDialogPreference.java:
```diff
@@ -223,8 +221,7 @@ private View configureLayoutForView(LinearLayout view) {
         if (shouldShowLocalNotifications) {
-            boolean isBloggingRemindersEnabled =
-                    mBloggingRemindersProvider != null && mBloggingRemindersProvider.isEnabled();
+            boolean isBloggingRemindersEnabled = mBloggingRemindersProvider != null;
```
We removed the`mBloggingRemindersProvider.isEnabled()` here, but should we ad other conditions, `buildConfigWrapper.isJetpackApp && site.hasCapabilityEditPosts`

ravishanker replied:
Notifications are removed on WP.  There's no need to check again here if it is JP I guess.

irfano replied:
Good point!

irfano reviewed (changes requested):
My review is still in progress but I have some comments:
- I think the description is wrong, it mentions `self-hosted sites not connected to Jetpack,` but actually, this PR only removes blogging reminders from WP, nothing related to self-hosted sites. If so, can you update the description properly?
- We checked `hasCapabilityEditPosts`, but it returns false for self-hosted sites even the site has capability. More context: p1688572548724379/1687890995.075909-slack-C0180B5PRJ4
- Reminders button in Site Settings crashes for self-hosted, not Jetpack-connected sites. Actually, the button shouldn't be there. It's a current bug on `trunk` but it doesn't crash on `trunk`. So maybe we can fix that bug here because it causes a new crash here. Wdyt?
 
[device-2023-07-06-164752.webm](https://github.com/wordpress-mobile/WordPress-Android/assets/2471769/b1a86b19-8060-4a4c-9443-cf43aaa28062)

ravishanker committed:
Remove reminders

Removes reminders item from General in Site Settings for self-hosted, not Jetpack connected sites

ravishanker committed:
Update BloggingRemindersManagerTest.kt

ravishanker committed:
Update SiteListItemBuilder.kt

Fix Me appearing both under Menu and on Bottom Nav

ravishanker commented:
> * I think the description is wrong, it mentions `self-hosted sites not connected to Jetpack,` but actually, this PR only removes blogging reminders from WP, nothing related to self-hosted sites. If so, can you update the description properly?

Good point.  Fixed the logic now to cater for self-hosted, non jetpack connected sites

> * We checked `hasCapabilityEditPosts`, but it returns false for self-hosted sites even the site has capability. More context: p1688572548724379/1687890995.075909-slack-C0180B5PRJ4

Removed `hasCapabilityEditPosts` condition check

> * Reminders button in Site Settings crashes for self-hosted, not Jetpack-connected sites. Actually, the button shouldn't be there. It's a current bug on `trunk` but it doesn't crash on `trunk`. So maybe we can fix that bug here because it causes a new crash here. Wdyt?

True, it shouldn't be there at all, removed.

irfano reviewed (changes requested):
Thanks for the changes Ravi!

Why did you add `site.isUsingWpComRestApi` condition? The reminders should be enabled for self-hosted sites too. 
You can check this internal discussion: p1688572319801989/1687890995.075909-slack-C0180B5PRJ4
And also check the updated [iOS PR](https://github.com/wordpress-mobile/WordPress-iOS/issues/20971) description that you mentioned. 

Also, the tests are failing. We ran them multiple times in this PR, I guess they are not flaky.

twstokes commented:
👋 @ravishanker! I just wanted to add (which hopefully adds clarity to things) that on iOS the [recent changes](https://github.com/wordpress-mobile/WordPress-iOS/pull/20970) just ensure that `isUserCapableOf(.EditPosts)` is now called when determining if we should show the Blogging Reminders prompt. It just so happens to be for technical reasons that self-hosted sites not connected to Jetpack will return `false` here, making the feature unavailable for those blogs.

I can't identify a business case where we don't want to support self-hosted sites not connected to Jetpack, it's just the client-side logic doesn't allow us to at the moment.

irfano commented:
> 👋 @ravishanker! I just wanted to add (which hopefully adds clarity to things) that on iOS the [recent changes](https://github.com/wordpress-mobile/WordPress-iOS/pull/20970) just ensure that `isUserCapableOf(.EditPosts)` is now called when determining if we should show the Blogging Reminders prompt. It just so happens to be for technical reasons that self-hosted sites not connected to Jetpack will return `false` here, making the feature unavailable for those blogs.
> 
> I can't identify a business case where we don't want to support self-hosted sites not connected to Jetpack, it's just the client-side logic doesn't allow us to at the moment.

@twstokes, let me restate what you said to ensure I understand correctly. We're disabling blogging reminders for self-hosted sites that are not connected to Jetpack because we are unable to technically check `isUserCapableOf(.EditPosts)`. Is that correct?

I have found another site property called **`isSelfHostedAdmin`**. We can utilize this condition and we can at least make blogging reminders enabled for self-hosted not Jetpack-connected users who are admin.

ravishanker committed:
Fix business logic

Show blogging reminders for all on Jetpack app

ravishanker committed:
Update BloggingRemindersViewModelTest.kt

remove extra line

ravishanker commented:
@irfano - Thanks for the clarification.  Updated PR title, and description to just Feature config removal, and Jetpack app only. 🙇

ravishanker committed:
Update BlockEditorTests.kt

Temporarily disabling a failing test.  It seems to work locally though

pachlava commented:
Hey @ravishanker 👋 

I see you've removed an assertion from `e2ePublishSimplePost`. I can see that it failed for WordPress only (where the reminder is removed).

My guess is that because the test still expected for the reminder to be shown (and waited for it for some time), the `Post Published` toast was shown and gone while the test was still waiting for the blogging reminders, resulting in the fail in `verifyPostPublished`. 

I _think_ executing [this line](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/BlockEditorPage.kt#L112) for JP only should resolve the issue.

twstokes commented:
> @twstokes, let me restate what you said to ensure I understand correctly. We're disabling blogging reminders for self-hosted sites that are not connected to Jetpack because we are unable to technically check `isUserCapableOf(.EditPosts)`. Is that correct?
> 
> I have found another site property called **`isSelfHostedAdmin`**. We can utilize this condition and we can at least make blogging reminders enabled for self-hosted not Jetpack-connected users who are admin.

Exactly @irfano - that's what way I understand it. Thanks!

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/items/listitem/SiteListItemBuilder.kt:
```diff
@@ -202,8 +202,7 @@ class SiteListItemBuilder @Inject constructor(
         return if ((!buildConfigWrapper.isJetpackApp &&
                     jetpackFeatureRemovalPhaseHelper.shouldRemoveJetpackFeatures() &&
-                    site.hasCapabilityManageOptions) ||
-            !siteUtilsWrapper.isAccessedViaWPComRest(site)
```
We deleted the` !siteUtilsWrapper.isAccessedViaWPComRest(site)` condition from `buildMeItemIfAvailable`. What does this PR make related to the me button?

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java:
```diff
@@ -1988,9 +1988,8 @@ private void removeNonSelfHostedPreferences() {
      */
     private void removeGeneralSettingsExceptBloggingReminders() {
         PreferenceGroup group = (PreferenceGroup) findPreference(getString(R.string.pref_key_site_general));
-        if (group != null && mBloggingRemindersPref != null) {
+        if (group != null) {
             group.removeAll();
-            group.addPreference(mBloggingRemindersPref);
         }
     }
```
After this change, `removeGeneralSettingsExceptBloggingReminders()` removes everything. I think it would be better to rename the function.

irfano reviewed (changes requested):
@ravishanker, can we make blogging reminders enabled for self-hosted admins, as I stated in this [comment](https://github.com/wordpress-mobile/WordPress-Android/pull/18739#issuecomment-1627053537). We can utilize the `isSelfHostedAdmin` parameter for this purpose.

I raised this concern because currently, a user with the "Subscriber" role on a self-hosted, non-Jetpack connected site can set a blogging reminder. However, they are unable to actually publish any posts.

ravishanker commented:
Looks like there are a few things mixed up here, and turned out to be something else.  Closing this for now, in favour of not introducing new issues.  Probably, need to raise separate PRs for removing feature config, limiting to JP, and handling self hosted sites
